labs/a4.md

```diff
@@ -0,0 +1,164 @@
+3. Configuring sshd 
+   * Another [comprehensive](https://www.digitalocean.com/community/tutorials/how-to-tune-your-ssh-daemon-configuration-on-a-linux-vps)
+     DigitalOcean guide to `/etc/ssh/sshd_config`
+   * [Fail2Ban](https://fail2ban.org) and [Archwiki](https://wiki.archlinux.org/index.php/fail2ban) configuration instructions
+4. On-host firewalls (ufw)
+   * Fairly comprehensive [ufw](https://help.ubuntu.com/lts/serverguide/firewall.html) configuration guide
+5. The Filesystem Hierarchy 
+   * [Wikipedia](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) entry
+   * Even more comprehensive [TLDP](https://www.tldp.org/LDP/Linux-Filesystem-Hierarchy/html/index.html) entry
+
+# Lab Activity
+
+In this lab you will implement some of the things we 
+discussed during lecture.
+
+### Generating and using SSH keys
+
+Passwords can be guessed, and this is bad. Good security practice now involves 
+taking advantage of public-key crypto to authenticate SSH sessions. In this
+section, we will be setting up SSH keys, if you haven't already.
+
+Please log into `tsunami.ocf.berkeley.edu` first. These instructions will not 
+work on a Windows machine.
+
+1. First, make sure you don't already have an existing SSH key in use that you 
+might overwrite:
+
+        you@tsunami $ ls -la ~/.ssh
+        -rw------- 1 <you> ocf <stuff> .ssh/id_rsa # if this already exists, continue to step 3
+
+2. If you don't have an SSH key, make one with the following command
+    
+        you@tsunami $ ssh-keygen -t rsa -b 4096
+    
+    You can give the key a passphrase if you'd like. This command creates a 
+    4096-bit RSA private key (`~/.ssh/id_rsa`) and corresponding public key
+    (`~/.ssh/id_rsa.pub`). There are many types of keys in various sizes, such
+    2048-bit RSA keys (`ssh-keygen -t rsa -b 2048`) or 384-bit ECDSA keys
+    (`ssh-keygen -t ecdsa -t 384`) or more. The full list can be found in
+    `man ssh-keygen`. Which one to use is up to your preference and/or
+    [paranoia][^paranoia] about quasi-legal government surveillance.
+
+3. Copy the public key over to your student VM
+
+        you@tsunami $ ssh-copy-id you@you.decal.xcf.sh
+    
+    This command, as its name suggests, copies your public key (by default, 
+    `~/.ssh/id_rsa.pub`) to the specified remote host, and appends the contents
+    of the public key to the `~/.ssh/authorized_keys` file, which is used to
+    control which keys are authorized to be used to authenticate a user. To see
+    specifically what happened, do `cat ~/.ssh/id_rsa.pub`) on `tsunami`
+    and then compare that to `cat ~/.ssh/authorized_keys` on your student VM.
+
+At the end `ssh-copy-id` will prompt you to attempt to log into the server 
+using your key. This should succeed. 
+
+_Checkoff: paste the contents of `~/.ssh/authorized_keys` from your 
+student VM._
+
+Checkoff bonus: configure `sshd` to only allow key-based login.
```

Comment:
what is a "checkoff bonus"?

Reply:
I should reword the "checkoff" wording, thanks.